### PR TITLE
[improvement] better debugging for concurrency leaks

### PR DESCRIPTION
**Pull request by j-baker**
This changes 3 things.

1. The limit log line now also contains the number of waiting requests.
2. The limit log line now also contains the hostname (unsafe, to be
consistent).
3. We keep a 'leak detector' which uses weak references to log and
optionally log the stack traces of places which we can know definitely
leaked.

### CODE REVIEW

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ConcurrencyLimiters.java [hunk 1]
```diff
@@ -204,42 +204,41 @@ public void onDropped() {}
...
             return future;
         }
 
         synchronized void processQueue() {
             while (!waitingRequests.isEmpty()) {
-                log.debug("Limit",
-                        SafeArg.of("limit", limiter.getLimit()),
-                        safeArgMethod,
-                        safeArgPathTemplate);
+                if (log.isDebugEnabled()) {
```
Comment (robert3005) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ConcurrencyLimiters.java:234]: is this debug line expensive?

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ConcurrencyLimiters.java [hunk 2]
```diff
@@ -204,42 +204,41 @@ public void onDropped() {}
...
-        private final ThreadWorkQueue<SettableFuture<Limiter.Listener>> waitingRequests = new ThreadWorkQueue<>();
+        private final ThreadWorkQueue<QueuedRequest> waitingRequests = new ThreadWorkQueue<>();
         @GuardedBy("this")
         private SimpleLimiter<Void> limiter;
         @GuardedBy("this")
         private ScheduledFuture<?> timeoutCleanup;
         private final Key limiterKey;
         private final Supplier<SimpleLimiter<Void>> limiterFactory;
-
-        private final SafeArg<Optional<String>> safeArgMethod;
```
Comment (robert3005) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ConcurrencyLimiters.java:217]: why the change?

File: okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/LeakDetectorTest.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.okhttp;
+
+import static org.assertj.core.api.Java6Assertions.assertThat;
```
Comment (robert3005) [inline okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/LeakDetectorTest.java]: I think you want to import Assertions class

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ConcurrencyLimiters.java (hunk 2 above)
Comment (j-baker) [reply okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ConcurrencyLimiters.java:217]: figure there's still two safeargs which will get made on ever call, figured it's easier to avoid them

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ConcurrencyLimiters.java (hunk 1 above)
Comment (j-baker) [reply okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ConcurrencyLimiters.java:234]: not hugely, but felt like... it's a debug line?

**Review by robert3005 — APPROVED** [review]

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/LeakDetector.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        if (stackTrace.isPresent()) {
+            log.warn("Resource leak detected - did you forget to close a resource properly? "
+                            + "This will likely hurt performance. Stack trace is of the call where "
+                            + "the acquire happened.",
+                    SafeArg.of("resourceType", resourceType.getName()),
+                    stackTrace.get());
+        } else {
+            log.warn("Leak detected in Conjure call - did you forget to close a resource properly? "
+                            + "This will likely hurt performance. To get a "
+                            + "stack trace for the call where the acquire happened, set log "
```
Comment (CRogers) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/LeakDetector.java:92]: What stack trace? :p
